### DISCUSSION TITLE
Support for Multiple Node Labels (derived from multiple RDF types)

### DISCUSSION THREAD

**Discussion by mnsRG** [kuzudb/kuzu#3114] [Ideas]
Hello, I was wondering if there is any plan to support multiple labels per Node. I am mainly dealing with RDF (and RDF-star) data and many nodes come with inferred types (eg a 'Desk' is also a 'Furniture'). Therefore, I would like to include all of them (not necessarily by keeping their hierarchies - flat is fine, which is how Neo4j handles it more or less). Then, when I ask for all the 'Furniture' in my database, I also get the 'Desk' nodes. 
Btw, I am not using the recommended way to load RDF data, because it is important that the schema of the database reflects its actual structure, as this provides the context for an LLM that translates natural language to cypher queries. 
Thanks!

**Comment by hpvd** [comment]
+1 on multiple labels for nodes and edges/relationships

could be pretty handy e.g. for
- seperating graph parts (subgraphs)
- help with sub indexes
- manage access (rely on label in separate app)
- tag with domain
- ..

**Comment by semihsalihoglu-uw** [comment]
Hi All,

Unfortunately, we don't have good ideas here, and this is not in our roadmap. I have always found the multi-label decision of the original property graphs not well-thought. If we were to support multiple labels, I would want to think about how to support a general type hierarchy. However, any system that supports general type hierarchies is losing in performance and we are trying not to sacrifice that. I have not seen a good approach (going back to object-oriented dbmss of 90s) that supported type hierarchies in a performant way. To set the expectations correctly though, even if we support that, it won't happen quickly.  

I think encoding the labels somehow as a List[String/INT] property is one manual approach of course but I understand that's far from ideal. If you have other manual approaches, it would be good to share with others in the community.

@mnsRG: When you say "recommended way to load RDF data", I assume you mean Kuzu's [RDFGraphs](https://docs.kuzudb.com/rdf-graphs/) feature.

**Reply by mnsRG** [reply]
@semihsalihoglu-uw yup I am referring to this way to load RDF. The current approach seems as a nice workaround to load 'any' rdf dataset (regardless of its implicit schema), but when it is important to express the structure of the data to the schema itself, then another way is needed. But this goes a bit off the multi-label topic (probably an alternative for loading RDF would be an intermediate step to read the implicit schema, write it to kuzudb and then ingest the data - plus find a way to deal with updates)

**Reply by semihsalihoglu-uw** [reply]
OK, RDFGraphs is one specific way of mapping RDF triples, so yes if you want the schema to expressed as labels, this won't work. I am not sure if we as a system can support multiple mappings from RDF or how easy that would be. We have to accumulate some feedback and suggestion on RDFGraphs before we make any decisions on this.

**Comment by hpvd** [comment]
maybe of interest, this is a topic not that uncommon, as one might think.
E.g. in falkordb (https://github.com/falkordb/falkordb -> formely redisgraph) there was the same request some years ago https://github.com/RedisGraph/RedisGraph/issues/13
and they seem to have found a way to make it possible without scarifying their good performance (different db type: server, in-memory)
not sure, but maybe its worth looking into it, even if they have an other architecture.

**Comment by hpvd** [comment]
just another idea on this, terminusdb (different graph db type: git like full history/immutable, no cypher) also support multiple labels in a somehow different way and is even with this feature-set faster than neo4j in several queries...

> In Neo4j we can add multi labels to a node so our node is Person and Doctor at the same time.
> In TerminusDB classes can be subclasses of other classes, so let’s add a subclass for scm:Person called scm:Doctor.

Creating hierarchies at https://terminusdb.com/blog/graph-databases-terminusdb-vs-neo4j/

**Reply by prrao87** [reply]
I've used TerminusDB before, and it's a very different DB with different goals than Kùzu :). It's a document graph DB that's very deep into RDF territory and the concept of inheritance and object-oriented models are ingrained into the solution. In Kùzu's case we actively want to bring the worlds of property graphs and RDF together via a unified query interface, Cypher. However, RDF graphs in Kùzu is still a wrapper around the underlying structured property graph model, which goes back to the earlier point Semih made about performance. Supporting columnar data structures and vectorized computations for speed, while managing object-oriented inheritance is a really challenging problem!

**Reply by hpvd** [reply]
thanks for bring up the details! My point is only, that `multi label` is pretty common in different kinds of graph dbs and it's there because there are many use cases for this.

**Reply by semihsalihoglu-uw** [reply]
Yes, this is the type of class hierarchy that makes more sense to me. But I am not aware of a principled system implementation that does this well. Systems can be faster or slower than each other for a variety of reasons, so I would be suspicious that terminusdb is faster for any reasons related to this class hierarchy implementation. I would be curious to learn of a principled design for supporting such general class hierarchies. Not saying there isn't but having perused oo dbms and gdbms literature for years, I'm unaware of one. If you have a many layered class hierarchy and you ask join queries with types that are in the higher levels of the hierarchy, you are forcing the system to consider many combinations of lower classes, which are always hard to evaluate.

**Reply by prrao87** [reply]
TerminusDB is fast mainly because it's implemented in Rust (Neo4j is all Java), and it also uses recursive operations on data structures really well in Rust. They've long been fans of the Prolog language and wanted to implement a solution that allows for efficient traversal of class hierarchies, hence their emphasis on RDF and OWL.

**Reply by prrao87** [reply]
What I'm curious about is the overlap between property graph models and RDF out in the literature or the wild. There aren't too many examples of tools I find that are trying to do what Kùzu is - blending the benefits of both worlds. That being said, there definitely are a lot of use cases in the real world for multilabel graphs and I agree with @hpvd's point there :).

**Reply by hpvd** [reply]
Regarding terminus db, they shared this imho interesting internal:
https://terminusdb.com/blog/terminusdb-internals/
https://terminusdb.com/blog/terminusdb-internals-2/

**Comment by hpvd** [comment]
what do you think of converting this to an issue, type feature request?

**Reply by hpvd** [reply]
new related issue created:
https://github.com/kuzudb/kuzu/issues/3117

**Reply by semihsalihoglu-uw** [reply]
Thanks for opening the issue. Let's keep it around and see if there is anything clever we can come up with at some point.

**Comment by hpvd** [comment]
if you are interested on this, plz comment /ad +1 in https://github.com/kuzudb/kuzu/issues/3117
if you have any ideas on use cases / requirements or performant implementation: plz share in https://github.com/kuzudb/kuzu/issues/3117

**Comment by angelo-hub** [comment]
+1 we want a generic contains
